src/Hintman/Installation.hs

```diff
@@ -84,7 +102,31 @@ createAccessToken installationId appInfo = do
     if status `elem` [200, 201]
         then do
             log I $ "Successfully created token from: " <> show installationId
-            pure $ decode body
+            case decode @AccessToken body of
+                Nothing -> throwError $ ServerError $
+                    "Error decoding token from: " <> decodeUtf8 body
+                Just AccessToken{..} -> pure InstallationToken
+                    { itToken = atToken
+                    , itExpiresAt = atExpiresAt
+                    , itInstallationId = installationId
+                    }
         else do
-            log E $ "Couldn't create token from: " <> show installationId
-            pure Nothing
+            log E $ "Couldn't create token for: " <> show installationId
+            throwError $ ServerError $
+                "Invalid status " <> show status <> " for " <> show installationId
+
+{- | This function takes 'InstallationAccessToken' and either returns itself if
+it can live long enough or creates a new token by calling 'createAccessToken'
+function. 'InstallationAccessToken' contains enough information to be renewed.
+-}
+renewToken
+    :: (MonadIO m, Has AppInfo env, WithError m, WithLog env m)
+    => InstallationToken
+    -> m InstallationToken
+renewToken token@InstallationToken{..} = do
+    now <- liftIO getCurrentTime
+
+    -- if less than 10 seconds left for token expiration, create new one
+    if now `diffUTCTime` itExpiresAt <= 10
```

Comment:
Isn't it milliseconds?

Reply:
Unfortunately, the API for time diffs is not clear enough and `fromInteger` from the `Num` instance converts from seconds. I double-checked the documentation:

* http://hackage.haskell.org/package/time-1.9.3/docs/Data-Time-Clock.html#t:NominalDiffTime